### PR TITLE
Added `databricks labs ucx move` command 

### PR DESCRIPTION
**Pull request by HariGS-DB**
closes https://github.com/databrickslabs/ucx/issues/675

### CODE REVIEW

File: labs.yml [hunk 1]
```diff
@@ -66,3 +66,17 @@ commands:
       - name: delete_managed
         description: Revert and delete managed tables
 
+
+  - name: migrate-uc-to-uc
```
Comment (nfx) [inline labs.yml:70]: command has to be `move`. that's how we agreed to call it.

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 2]
```diff
@@ -194,3 +196,103 @@ def _get_mapping_rules(self) -> dict[str, Rule]:
...
+                    if self._validate_uc_objects("table", f"{to_catalog}.{to_schema}.{table.name}") == 1:
+                        logger.warning(
+                            f"table {from_table} already present in {from_catalog}.{from_schema}."
+                            f" skipping this table..."
+                        )
+                        continue
+                    if table.table_type and table.table_type.value in ("EXTERNAL", "MANAGED"):
+                        table_tasks.append(
+                            partial(
+                                self._migrate_uc_table, from_catalog, from_schema, table.name, to_catalog, to_schema
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:226]: too much nesting and it is unreadable. refactor.

**Review by nfx — CHANGES_REQUESTED** [review]
please review the parallel PR not to do a duplicate work - https://github.com/databrickslabs/ucx/pull/755

File: src/databricks/labs/ucx/cli.py [hunk 3]
```diff
@@ -149,6 +149,47 @@ def revert_migrated_tables(schema: str, table: str, *, delete_managed: bool = Fa
...
+        logger.error("Please enter from_catalog and to_catalog details")
+        return
+    if from_schema == "" or to_schema == "" or from_table == "":
+        logger.error("Please enter from_schema, to_schema and from_table(enter * for migrating all tables) details.")
+        return
+    if from_catalog == to_catalog and from_schema == to_schema:
+        logger.error("please select a different schema or catalog to migrate to")
+        return
+    logger.error(f"migrating tables {from_table} from {from_catalog}.{from_schema} to {to_catalog}.{to_schema}")
+    tables.migrate_uc_tables(
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:184]: If you want to support wildcard, check for "*" value and iterate over tables from that schema and call this method

**Review by nfx — CHANGES_REQUESTED** [review]
Code is incorrect and has bugs

File: labs.yml (hunk 1 above)
Comment (nfx) [reply labs.yml:70]: Command still has previous name. Fix it.

File: src/databricks/labs/ucx/cli.py [hunk 4]
```diff
@@ -165,5 +165,48 @@ def revert_migrated_tables(w: WorkspaceClient, schema: str, table: str, *, delet
...
+        logger.error("Please enter from_catalog and to_catalog details")
+        return
+    if from_schema == "" or to_schema == "" or from_table == "":
+        logger.error("Please enter from_schema, to_schema and from_table(enter * for migrating all tables) details.")
+        return
+    if from_catalog == to_catalog and from_schema == to_schema:
+        logger.error("please select a different schema or catalog to migrate to")
+        return
+    logger.error(f"migrating tables {from_table} from {from_catalog}.{from_schema} to {to_catalog}.{to_schema}")
+    tables.move_migrated_tables(
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:202]: What if tables are not migrated?

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 5]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+    def _get_mapping_rules(self) -> dict[str, Rule]:
+        mapping_rules: dict[str, Rule] = {}
+        for rule in self._tm.load():
+            mapping_rules[rule.as_hms_table_key] = rule
+        return mapping_rules
+
+    def move_migrated_tables(
+        self, from_catalog: str, from_schema: str, from_table: str, to_catalog: str, to_schema: str
+    ):
+        self._ws.schemas.get(f"{from_catalog}.{from_schema}")
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:200]: What this line is doing? You're not using a result. It's redundant

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 6]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+        filtered_tables = [table for table in tables if from_table in [table.name, "*"]]
+        for table in filtered_tables:
+            try:
+                self._ws.tables.get(full_name=f"{to_catalog}.{to_schema}.{table.name}")
+                logger.warning(
+                    f"table {from_table} already present in {from_catalog}.{from_schema}. skipping this table..."
+                )
+                continue
+            except NotFound:
+                if table.table_type and table.table_type.value in ("EXTERNAL", "MANAGED"):
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:219]: Use enum, eg TableType.EXTERNAL, not the string

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 7]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+                    table_tasks.append(
+                        partial(self._migrate_uc_table, from_catalog, from_schema, table.name, to_catalog, to_schema)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:221]: This name looks misleading. Has to be "_move_table"

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 8]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+                    )
+                else:
+                    view_tasks.append(
+                        partial(
+                            self._migrate_uc_view,
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:226]:
```suggestion
                            self._move_view,
```

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 9]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
             print("To revert and delete Migrated Tables, add --delete_managed true flag to the command.")
         return True
+
+    def _get_mapping_rules(self) -> dict[str, Rule]:
+        mapping_rules: dict[str, Rule] = {}
+        for rule in self._tm.load():
+            mapping_rules[rule.as_hms_table_key] = rule
+        return mapping_rules
+
+    def move_migrated_tables(
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:197]: Can you actually create a separate class for this functionality

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 10]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
             print("Migrated Manged Tables (targets) will be left intact.")
             print("To revert and delete Migrated Tables, add --delete_managed true flag to the command.")
         return True
+
+    def _get_mapping_rules(self) -> dict[str, Rule]:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:191]: This change looks out of scope

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 11]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+        from_schema: str,
+        from_table: str,
+        to_catalog: str,
+        to_schema: str,
+    ) -> bool:
+        from_table_name = f"{from_catalog}.{from_schema}.{from_table}"
+        to_table_name = f"{to_catalog}.{to_schema}.{from_table}"
+        try:
+            create_sql = str(next(self._backend.fetch(f"SHOW CREATE TABLE {from_table_name}"))[0])
+            create_sql = create_sql.replace(from_catalog, to_catalog).replace(from_schema, to_schema)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:252]: Bug: what if some of the columns have the same name as catalog/schema? It's going to break the data

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/cli.py (hunk 4 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/cli.py:202]: @nfx  The calling method tables.move_migrated_tables should throw the relevant error messages and info of tables migrated. should we need to add a try catch here too in the cli?

File: labs.yml (hunk 1 above)
Comment (nfx) [reply labs.yml:70]: Repeating for the third time: this command has to be called "move"

File: src/databricks/labs/ucx/cli.py [hunk 12]
```diff
@@ -165,5 +165,48 @@ def revert_migrated_tables(w: WorkspaceClient, schema: str, table: str, *, delet
         tm.revert_migrated_tables(schema, table, delete_managed=delete_managed)
 
 
+@ucx.command
+def migrate_uc_to_uc(
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:169]:
```suggestion
def move(
```

Repeating for the third time: this command has to be called "move"

File: src/databricks/labs/ucx/cli.py [hunk 13]
```diff
@@ -165,5 +165,48 @@ def revert_migrated_tables(w: WorkspaceClient, schema: str, table: str, *, delet
...
+    logger.info("Running move command")
+    installation_manager = InstallationManager(w)
+    installation = installation_manager.for_user(w.current_user.me())
+    if not installation:
+        logger.error(CANT_FIND_UCX_MSG)
+        return
+    sql_backend = StatementExecutionBackend(w, installation.config.warehouse_id)
+    tmp = TableMapping(w, sql_backend)
+    tables = TablesMigrate(
+        TablesCrawler(sql_backend, installation.config.inventory_database),
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:187]: Can you extract the variable for table crawler, so that the surrounding constructor call doesn't look odd?

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py (hunk 10 above)
Comment (nfx) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:191]: Repeating for the second time: these changes are out of scope for this pr

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py (hunk 9 above)
Comment (nfx) [reply src/databricks/labs/ucx/hive_metastore/table_migrate.py:197]: Please introduce a separate class for uc2uc, it'll be easier this way

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 14]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+                            from_catalog,
+                            from_schema,
+                            table.name,
+                            to_catalog,
+                            to_schema,
+                            table.view_definition,
+                        )
+                    )
+        Threads.strict(name="creating tables", tasks=table_tasks)
+        logger.info(f"migrated {len(list(table_tasks))} tables to the new schema {to_schema}.")
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:236]:
```suggestion
        logger.info(f"moved {len(list(table_tasks))} tables to the new schema {to_schema}.")
```

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 15]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+                            self._migrate_uc_view,
+                            from_catalog,
+                            from_schema,
+                            table.name,
+                            to_catalog,
+                            to_schema,
+                            table.view_definition,
+                        )
+                    )
+        Threads.strict(name="creating tables", tasks=table_tasks)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:235]: Don't use named arguments if they're positional. It's harder to read

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 16]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+            grants = self._ws.grants.get(securable_type=SecurableType.TABLE, full_name=from_table_name)
+            if grants.privilege_assignments is None:
+                return True
+            grants_changes = [
+                PermissionsChange(add=pair.privileges, principal=pair.principal)
+                for pair in grants.privilege_assignments
+            ]
+            self._ws.grants.update(securable_type=SecurableType.TABLE, full_name=to_table_name, changes=grants_changes)
+            return True
+        except DatabricksError:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:264]: This error is too generic and hides the real problems with the code. You are allowed to catch only subclasses of it. The other errors will bubble up

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 17]
```diff
@@ -184,3 +187,108 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:292]: This error is too broad. Catch more concrete errors

**Review by nfx — CHANGES_REQUESTED** [review]
Don't mark comments as resolved without resolving them

File: src/databricks/labs/ucx/cli.py [hunk 18]
```diff
@@ -165,5 +165,41 @@ def revert_migrated_tables(w: WorkspaceClient, schema: str, table: str, *, delet
...
+    installation_manager = InstallationManager(w)
+    installation = installation_manager.for_user(w.current_user.me())
+    if not installation:
+        logger.error(CANT_FIND_UCX_MSG)
+        return
+    sql_backend = StatementExecutionBackend(w, installation.config.warehouse_id)
+    tables = TableMove(
+        w,
+        sql_backend,
+    )
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:189]:
```suggestion
    tables = TableMove(w, sql_backend)
```

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 19]
```diff
@@ -182,3 +184,144 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
             print("To revert and delete Migrated Tables, add --delete_managed true flag to the command.")
         return True
+
+
+class TableMove:
+    def __init__(
+        self,
+        ws: WorkspaceClient,
+        backend: SqlBackend,
+    ):
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:194]:
```suggestion
    def __init__(self, ws: WorkspaceClient, backend: SqlBackend):
```

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 20]
```diff
@@ -182,3 +184,144 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+            grants_changes = [
+                PermissionsChange(pair.privileges, pair.principal) for pair in grants.privilege_assignments
+            ]
+            self._ws.grants.update(SecurableType.TABLE, to_table_name, changes=grants_changes)
+            if del_table:
+                logger.info(f"dropping source table {from_table_name}")
+                drop_sql = f"DROP TABLE {from_table_name}"
+                self._backend.execute(drop_sql)
+            return True
+
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:282]:
```suggestion
```

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 21]
```diff
@@ -182,3 +184,144 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+        except NotFound as err:
+            if "[TABLE_OR_VIEW_NOT_FOUND]" in str(err):
+                logger.error(f"Could not find table {from_table_name}. Table not found.")
+            else:
+                logger.error(err)
+        except PermissionDenied:
+            logger.error(f"error applying permissions for table {to_table_name}")
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:289]:
```suggestion
```
you shouldn't swallow permission errors.

File: src/databricks/labs/ucx/hive_metastore/table_migrate.py [hunk 22]
```diff
@@ -182,3 +184,144 @@ def print_revert_report(self, *, delete_managed: bool) -> bool | None:
...
+                drop_sql = f"DROP VIEW {from_table_name}"
+                self._backend.execute(drop_sql)
+            return True
+        except NotFound as err:
+            if "[TABLE_OR_VIEW_NOT_FOUND]" in str(err):
+                logger.error(f"Could not find view {from_table_name}. View not found.")
+            else:
+                logger.error(err)
+        except PermissionDenied:
+            logger.error(f"error applying permissions for view {to_table_name}")
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_migrate.py:326]:
```suggestion
```
don't swallow permission errors

File: tests/integration/hive_metastore/test_table_move.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+    assert (
+        len(
+            [
+                rec.message
+                for rec in caplog.records
+                if f"schema {from_schema} not found in catalog {from_catalog.name}" in rec.message
+            ]
+        )
+        == 1
+    )
```
Comment (nfx) [inline tests/integration/hive_metastore/test_table_move.py:29]: this is unreadable, please refactor

File: tests/integration/hive_metastore/test_table_move.py [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+    from_table_1 = make_table(catalog_name=from_catalog.name, schema_name=from_schema.name)
+    from_table_2 = make_table(catalog_name=from_catalog.name, schema_name=from_schema.name)
+    from_table_3 = make_table(catalog_name=from_catalog.name, schema_name=from_schema.name)
+    to_catalog = make_catalog()
+    to_schema = make_random(4)
+    tm.move_tables(from_catalog.name, from_schema.name, from_table_1.name, to_catalog.name, to_schema, True)
+    tables = ws.tables.list(catalog_name=to_catalog.name, schema_name=to_schema)
+    dropped_tables = ws.tables.list(catalog_name=from_catalog.name, schema_name=from_schema.name)
+    assert len([t for t in tables if t.name in [from_table_1.name, from_table_2.name, from_table_3.name]]) == 1
+    assert len(list(dropped_tables)) == 3
```
Comment (nfx) [inline tests/integration/hive_metastore/test_table_move.py:103]: please assert on exact names of tables that were dropped

File: tests/integration/hive_metastore/test_table_move.py [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+    )
+    view_1_grant = ws.grants.get(
+        securable_type=SecurableType.TABLE, full_name=f"{to_catalog.name}.{to_schema.name}.{from_view_1.name}"
+    )
+    assert (
+        len(
+            [t for t in tables if t.name in [from_table_1.name, from_table_2.name, from_table_3.name, from_view_1.name]]
+        )
+        == 4
+    )
```
Comment (nfx) [inline tests/integration/hive_metastore/test_table_move.py:75]: this is unreadable, please refactor. don't assert for length - use "assert key in dict" instead

File: tests/integration/hive_metastore/test_table_move.py [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+    expected_table_1_grant = [PrivilegeAssignment(group_a.display_name, [Privilege.SELECT])]
+    expected_table_2_grant = [
+        PrivilegeAssignment(group_b.display_name, [Privilege.MODIFY, Privilege.SELECT]),
+    ]
+    expected_table_3_grant = [PrivilegeAssignment(group_a.display_name, [Privilege.SELECT])]
+    expected_view_1_grant = [PrivilegeAssignment(group_b.display_name, [Privilege.SELECT])]
+
+    assert len([perm for perm in table_1_grant.privilege_assignments if perm in expected_table_1_grant]) == 1
```
Comment (nfx) [inline tests/integration/hive_metastore/test_table_move.py:83]: don't assert for length - use "assert key in dict" instead

**Review by nfx — CHANGES_REQUESTED** [review]

**Review by nfx — APPROVED** [review]
lgtm
